src/company_gui.cpp

```diff
@@ -1827,267 +1841,221 @@ struct CompanyInfrastructureWindow : Window
 		this->roadtypes &= ~_roadtypes_hidden_mask;
 	}
 
-	/** Get total infrastructure maintenance cost. */
-	Money GetTotalMaintenanceCost() const
+	/**
+	 * Build a vector of infrastructure types(roads, trains, trams, airports, stations, water vehicles) to be displayed.
+	 *   Total maintenance cost of all infrastructure types is last entry in vector.
+	 */
+	void BuildCompanyInfrastructureList() const
 	{
 		const Company *c = Company::Get((CompanyID)this->window_number);
-		Money total;
 
-		uint32_t rail_total = c->infrastructure.GetRailTotal();
-		for (RailType rt = RAILTYPE_BEGIN; rt != RAILTYPE_END; rt++) {
-			if (HasBit(this->railtypes, rt)) total += RailMaintenanceCost(rt, c->infrastructure.rail[rt], rail_total);
-		}
-		total += SignalMaintenanceCost(c->infrastructure.signal);
+		Money maintenance_cost = 0;
+		Money total_maintenance_cost = 0;
+
+
+		infrastructurelist.clear();
+
+
+		/* railtypes + signals. */
+		infrastructurelist.push_back(CIWInfrastructureCache(CIW_ENTRY_HEADER, STR_COMPANY_INFRASTRUCTURE_VIEW_RAIL_SECT));
```

Comment:
Set up constructors for value-less and value-ful rows.

Then you can do `infrastructurelist.emplace_back(CIW_ENTRY_HEADER, STR_COMPANY_INFRASTRUCTURE_VIEW_RAIL_SECT);` etc for all these lines. No need to keep listing the type that the container contains.

Reply:
Thank you for the tip.

Let me review my commit.

Reply:
updated.

Thank you.